### ISSUE TITLE
[BUG] Static Covariates takes insane amount of time in scaling

### ISSUE THREAD

**Issue by TamerAbdelmigid** [unit8co/darts#1041]
**Describe the bug**
I don't know if it is a bug or I am doing something wrong, but static covariates takes 42 min just to scale

**To Reproduce**
```
os.chdir("C:\\Users\\Tamer\\Documents\\Python Scripts\\PhD\\Data\\Estimator Forecaster\\")
cwd = os.getcwd()

df = pd.read_parquet("Forecaster_100R.parquet")
df.index = pd.RangeIndex(start=0, stop=len(df), step=1)
static_covariates = pd.Series([e for e in [1,2,3,4,5,6] for i in range(int(len(df)/6))])
X = df.iloc[:, -1]
X.index = pd.RangeIndex(start=0, stop=len(X), step=1)
```
```
%%time
series = TimeSeries.from_series(X, static_covariates=static_covariates)

```
```
CPU times: total: 4min 13s
Wall time: 4min 56s
```
```
%%time
scaler = Scaler()

idx = int(len(series) - 10000)
train, val = series[:idx], series[idx:]
train_scaled = scaler.fit_transform(train).astype(np.float32)
val_scaled = scaler.transform(val).astype(np.float32)
series_scaled = train_scaled.concatenate(val_scaled).astype(np.float32)

wdir = "H:\\PhD\\Expirements\\exp3\\"
os.chdir(wdir)
cwd1 = os.getcwd()
logging.basicConfig(level=logging.DEBUG)
```
```
CPU times: total: 31min 53s
Wall time: 41min 22s
```

**Expected behavior**
It shouldn't take this much time, by repeating the same code without the static covariates, it takes 13ms, and 53ms respectively.

**System (please complete the following information):**
 - Python version: [e.g. 3.9]
 - darts version [e.g. 0.20.0]
 - CPU: AMD Ryzen 3950X

**Additional context**
`len(X)=2400600`

**Comment by gdevos010** [comment]
I can't speak to if this is a bug or not but you can make `Scaler` multithreaded by passing `n_jobs=-1`.

**Comment by TamerAbdelmigid** [comment]
Hello, @gdevos010, thanks for your advice, really appreciated. 
Yet, multithreaded scaling works if there are several timeseries, but my case is univariate, I tried it anyway and it made no difference, as for scaling before, it is a good idea, but still, dealing with static covariates shouldn't take that long.

**Comment by hrzn** [comment]
Normally the static covariates should not be affected by (or, I believe, affecting) scaling. We should investigate and profile this. Could you by any chance provide a self-contained minimal code snippet we could use for replication?

**Comment by TamerAbdelmigid** [comment]
Thanks @hrzn for replying. I have generated minimal self contained code below that replicates the problem as you requested.

```
import numpy as np
import pandas as pd
from darts import TimeSeries
from darts.dataprocessing.transformers import Scaler

X = pd.DataFrame(np.random.uniform(-5,5,2400600))
static_covariates = pd.Series([e for e in [1,2,3,4,5,6] for i in range(int(len(X)/6))])

series = TimeSeries.from_series(X, static_covariates=static_covariates)

scaler = Scaler()
train, val = series[:10000], series[10000:]
train_scaled = scaler.fit_transform(train).astype(np.float32)
val_scaled = scaler.transform(val).astype(np.float32)
series_scaled = train_scaled.concatenate(val_scaled).astype(np.float32)
```

**Comment by hrzn** [comment]
Thanks!
This overhead is due to casting of all the covariates columns to match the `TimeSeries`'s `dtype`. This casting is done at each `TimeSeries` creation (incl. when scaling but not only), which is causing the overhead. I've opened a PR to avoid the casting when that's not necessary: https://github.com/unit8co/darts/pull/1053

NB: Are you sure that you intended to create a static covariate of size 2400600? That's a lot of dimensions for a vector! Also I find it surprising that the static covariates size in your case has same dimensionality as the number of time steps of your series.

**Comment by TamerAbdelmigid** [comment]
@hrzn, I think by your question, I did not use static covariates correctly?!
Look, I have a univariate series composed of six sea state spectrums (i.e. wave height elevation for different sea conditions), lumped together they create 2400600 observations/samples, but I tried to use static covariates to differentiate which belong to sea state 1, and which belong to sea state 2 and so on. 
So knowing that, how should I implement static covariates? could you kindly elaborate?
Thanks in advance.

**Comment by hrzn** [comment]
Hi, from your description, it looks to me that you may be in one of the following situation:
* If your 6 state spectrum follow each other in time (i.e., state 1 is measured from t0 to t1, then state 2 from t1 to t2, etc), you can indeed represent them as one univariate series, which sort of contains 6 different regimes over time. Since the regime (or state) is time-varying, I would say in this case you should use a covariates time series (past or future, depending whether the "state" is known in advance at prediction time) 
* Or, you are talking about 6 distinct "observations" of the same quantity (wave height) in different sates. In that case, you probably should represent these as 6 distinct univariate `TimeSeries` (another option could be one multivariate `TimeSeries` with 6 components, if the 6 measurements have all been obtained simultaneously - i.e., share the same time stamps). In this case, you can attach to each of these 6 univariate series a static covariate representing the categorical "state" it captures. For instance using 1-hot encoding (and so a vector of size 6).

So indeed I strongly doubt you intended to use static covariates with one dimension per time step (since static covariates are static, it makes sense they don't vary with time).
I hope this helps.

**Comment by TamerAbdelmigid** [comment]
Thanks a lot @hrzn for you response, and sorry for taking some time to respond but I discovered I was implementing it wrong.

I have 100 experiments, each generating 4001 samples, this is repeated for 6 different sea states conditions (totaling 2400600). I was wrong in  stacking all the values along the time axis, I know now I should have used several time series in a sequence, as indicated in your [Meta Learning Guide](https://unit8co.github.io/darts/examples/14-transfer-learning.html) and previous response. 

But, now I am confused on how to implement the static covariates in the second example in your previous response.
Indeed, I will require only a vector of length 6. My current data shape is:
`sequence [SS1_exp1, SS1_exp2, ...., SS6_exp100]`
where SS1 is sea state 1, and exp1 is experiment 1, respectively. Hence, it is a list of 600 univariate timeseries, each contains 4001 observations, all share the same time axis. 

So, how to use the static covariates to differentiate which is sea state 1 and which is sea state 2 (I have searched for a guide for static covariates usage but haven't found one, except a small one in the method itself, and an [issue 1016](https://github.com/unit8co/darts/issues/1016) asking for a guide).

Can you please assist me with a sample code for my case to implement static covariates
